Userland/Libraries/LibGfx/PaintStyle.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2023, MacDue <macdue@dueutil.tech>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/QuickSort.h>
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/Vector.h>
+#include <LibGfx/Color.h>
+#include <LibGfx/Forward.h>
+#include <LibGfx/Gradients.h>
+#include <LibGfx/Rect.h>
+
+namespace Gfx {
+
+struct PaintStyle : RefCounted<PaintStyle> {
+    virtual ~PaintStyle() = default;
+    using SamplerFunction = Function<Color(IntPoint)>;
+    using PaintFunction = Function<void(SamplerFunction)>;
+
+    friend Painter;
+    friend AntiAliasingPainter;
+
+private:
+    // Simple paint styles can simply override sample_color() if they can easily generate a color from a coordinate.
+    virtual Color sample_color(IntPoint) const { return Color(); };
+
+    // Paint styles that have paint time dependent state (e.g. based on the paint size) may find it easier to override paint().
+    // If paint() is overridden sample_color() is unused.
+    virtual void paint(IntRect physical_bounding_box, PaintFunction paint) const
+    {
+        (void)physical_bounding_box;
+        paint([this](IntPoint point) { return sample_color(point); });
+    }
+};
+
+struct SolidPaintStyle : PaintStyle {
```

Comment:
Maybe `SolidColorPaintStyle` would be even clearer?
Also please mark this `final`.